File: clients/ruby/fiducia.rb
# Fiducia HTTP client (Ruby). Zero-dependency — stdlib net/http + json.
# Implements PROTOCOL.md.
#
#   require_relative "fiducia"
#   c = Fiducia::Client.new("https://api.fiducia.cloud")
#   lock = c.lock_acquire("orders/checkout", ttl_ms: 30000)
#   c.lock_release("orders/checkout", lock["result"]["lock_id"])

require "net/http"
require "json"
require "securerandom"
require "uri"

module Fiducia
  class Error < StandardError
    attr_reader :status, :body
    def initialize(status, body)
      super("fiducia: HTTP #{status}")
      @status = status
      @body = body
    end
  end

  # Raised by the blocking +lock+/+acquire_semaphore+ when the wait budget elapses.
  class LockTimeout < StandardError
    attr_reader :keys, :waited_ms
    def initialize(keys, waited_ms)
      super("fiducia: timed out after #{waited_ms}ms waiting for #{keys.join(', ')}")
      @keys = keys
      @waited_ms = waited_ms
    end
  end

  # A held lock grant. Call +release+ (alias +unlock+) when done.
  Lock = Struct.new(:client, :keys, :holder, :fencing_token, :lease_expires_ms) do
    def release; client.lock_release(holder, fencing_token); end
    alias_method :unlock, :release
  end

  # A held semaphore permit. Call +release+ when done.
  SemaphoreHandle = Struct.new(:client, :key, :holder, :fencing_token, :lease_expires_ms) do
    def release; client.semaphore_release(key, holder, fencing_token); end
    alias_method :unlock, :release
  end

  class Client
    def initialize(base_url)
      @base = base_url.sub(%r{/+\z}, "")
    end

    # --- misc ---
    def health; request("GET", "/healthz"); end
    def status; request("GET", "/v1/status"); end

    # --- locks (current protocol: holder + fencing_token, key in the body) ---
    def lock_get(key); request("GET", "/v1/locks?key=#{enc key}"); end
    def lock_acquire(keys, holder: nil, ttl_ms: nil, wait: false)
      request("POST", "/v1/locks/acquire", { keys: Array(keys), holder: holder, ttl_ms: ttl_ms, wait: wait })
    end
    def lock_release(holder, fencing_token)
      request("POST", "/v1/locks/release", { holder: holder, fencing_token: fencing_token })
    end

    # --- semaphores ---
    def semaphore_get(key); request("GET", "/v1/semaphores?key=#{enc key}"); end
    def semaphore_acquire(key, limit, holder: nil, ttl_ms: nil, wait: false)
      request("POST", "/v1/semaphores/acquire", { key: key, limit: limit, holder: holder, ttl_ms: ttl_ms, wait: wait })
    end
    def semaphore_release(key, holder, fencing_token)
      request("POST", "/v1/semaphores/release", { key: key, holder: holder, fencing_token: fencing_token })
    end

    # --- high-level blocking / try acquisition (live-mutex style) ---
    #
    # try_lock: wait:false — returns a Lock if free right now, else nil.
    # lock / must_lock: wait:true — blocks (polling) until acquired, the budget
    #   elapses (LockTimeout), or the server errors.
    def try_lock(key, ttl_ms: 60_000, holder: nil)
      _acquire_lock(Array(key), false, ttl_ms, holder, 0, 0, nil)
    end

    def lock(key, ttl_ms: 60_000, holder: nil, max_wait_ms: 30_000, retry_interval_ms: 250, max_retries: nil)
      got = _acquire_lock(Array(key), true, ttl_ms, holder, max_wait_ms, retry_interval_ms, max_retries)
      raise LockTimeout.new(Array(key), max_wait_ms) unless got
      got
    end
    alias_method :must_lock, :lock

    # Acquire, yield the Lock, then always release.
    def with_lock(key, **opts)
      held = lock(key, **opts)
      begin
        yield held
      ensure
        begin; held.release; rescue StandardError; end
      end
    end

    def try_semaphore(key, limit, ttl_ms: 60_000, holder: nil)
      _acquire_semaphore(key, limit, false, ttl_ms, holder, 0, 0, nil)
    end

    def acquire_semaphore(key, limit, ttl_ms: 60_000, holder: nil, max_wait_ms: 30_000, retry_interval_ms: 250, max_retries: nil)
      got = _acquire_semaphore(key, limit, true, ttl_ms, holder, max_wait_ms, retry_interval_ms, max_retries)
      raise LockTimeout.new([key], max_wait_ms) unless got
      got
    end

    # --- reader-writer locks ---
    def rw_acquire_read(key, ttl_ms: nil, wait: true)
      request("POST", "/v1/rw/#{enc key}/read", { ttl_ms: ttl_ms, wait: wait })
    end
    def rw_end_read(key, lock_id)
      request("POST", "/v1/rw/#{enc key}/read/end", { lock_id: lock_id })
    end
    def rw_acquire_write(key, ttl_ms: nil, wait: true)
      request("POST", "/v1/rw/#{enc key}/write", { ttl_ms: ttl_ms, wait: wait })
    end
    def rw_end_write(key, lock_id)
      request("POST", "/v1/rw/#{enc key}/write/end", { lock_id: lock_id })
    end

    # --- config KV ---
    def kv_get(key); request("GET", "/v1/kv/#{enc key}"); end
    def kv_put(key, value, ttl_ms: nil)
      request("PUT", "/v1/kv/#{enc key}", { value: value, ttl_ms: ttl_ms })
    end
    def kv_delete(key); request("DELETE", "/v1/kv/#{enc key}"); end
    def kv_list(prefix); request("GET", "/v1/kv?prefix=#{enc prefix}"); end

    # --- leader election ---
    def election_campaign(name, candidate, ttl_ms)
      request("POST", "/v1/elections/#{enc name}/campaign", { candidate: candidate, ttl_ms: ttl_ms })
    end
    def election_renew(name, candidate, fencing_token)
      request("POST", "/v1/elections/#{enc name}/renew", { candidate: candidate, fencing_token: fencing_token })
    end
    def election_resign(name, candidate, fencing_token)
      request("POST", "/v1/elections/#{enc name}/resign", { candidate: candidate, fencing_token: fencing_token })
    end
    def election_get(name); request("GET", "/v1/elections/#{enc name}"); end

    # --- service discovery ---
    def service_register(service, instance_id, address, ttl_ms)
      request("PUT", "/v1/services/#{enc service}/instances/#{enc instance_id}", { address: address, ttl_ms: ttl_ms })
    end
    def service_heartbeat(service, instance_id)
      request("POST", "/v1/services/#{enc service}/instances/#{enc instance_id}/heartbeat")
    end
    def service_deregister(service, instance_id)
      request("DELETE", "/v1/services/#{enc service}/instances/#{enc instance_id}")
    end
    def service_instances(service); request("GET", "/v1/services/#{enc service}"); end
    def service_list; request("GET", "/v1/services"); end

    private

    def enc(s)
      URI.encode_www_form_component(s.to_s)
    end

    def request(method, path, body = nil)
      uri = URI(@base + path)
      klass = {
        "GET" => Net::HTTP::Get, "PUT" => Net::HTTP::Put,
        "POST" => Net::HTTP::Post, "DELETE" => Net::HTTP::Delete
      }.fetch(method)
      req = klass.new(uri)
      if body
        req["content-type"] = "application/json"
        req.body = JSON.generate(body)
      end
      res = Net::HTTP.start(uri.host, uri.port, use_ssl: uri.scheme == "https") { |h| h.request(req) }
      data = res.body && !res.body.empty? ? JSON.parse(res.body) : nil
      raise Error.new(res.code.to_i, data) if res.code.to_i >= 300
      data
    end
  end
end

File: clients/php/Fiducia.php
<?php
// Fiducia HTTP client (PHP). Zero-dependency — ext-curl + json. Implements PROTOCOL.md.
//
//   require "Fiducia.php";
//   $c = new Fiducia\Client("https://api.fiducia.cloud");
//   $lock = $c->lockAcquire("orders/checkout", 30000);
//   $c->lockRelease("orders/checkout", $lock["result"]["lock_id"]);

namespace Fiducia;

class FiduciaException extends \Exception
{
    public int $status;
    public $body;
    public function __construct(int $status, $body)
    {
        parent::__construct("fiducia: HTTP $status");
        $this->status = $status;
        $this->body = $body;
    }
}

/** Thrown by the blocking lock()/acquireSemaphore() when the wait budget elapses. */
class LockTimeout extends \Exception
{
    public array $keys;
    public int $waitedMs;
    public function __construct(array $keys, int $waitedMs)
    {
        parent::__construct("fiducia: timed out after {$waitedMs}ms waiting for " . implode(", ", $keys));
        $this->keys = $keys;
        $this->waitedMs = $waitedMs;
    }
}

/** A held lock grant. Call release() (alias unlock()) when done. */
class Lock
{
    public function __construct(
        private Client $client,
        public array $keys,
        public string $holder,
        public int $fencingToken,
        public $leaseExpiresMs = null
    ) {
    }
    public function release()
    {
        return $this->client->lockRelease($this->holder, $this->fencingToken);
    }
    public function unlock()
    {
        return $this->release();
    }
}

/** A held semaphore permit. Call release() when done. */
class SemaphoreHandle
{
    public function __construct(
        private Client $client,
        public string $key,
        public string $holder,
        public int $fencingToken,
        public $leaseExpiresMs = null
    ) {
    }
    public function release()
    {
        return $this->client->semaphoreRelease($this->key, $this->holder, $this->fencingToken);
    }
    public function unlock()
    {
        return $this->release();
    }
}

class Client
{
    private string $base;

    public function __construct(string $baseUrl)
    {
        $this->base = rtrim($baseUrl, "/");
    }

    private function request(string $method, string $path, $body = null)
    {
        $ch = curl_init($this->base . $path);
        curl_setopt($ch, CURLOPT_CUSTOMREQUEST, $method);
        curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);
        if ($body !== null) {
            curl_setopt($ch, CURLOPT_POSTFIELDS, json_encode($body));
            curl_setopt($ch, CURLOPT_HTTPHEADER, ["content-type: application/json"]);
        }
        $resp = curl_exec($ch);
        $code = curl_getinfo($ch, CURLINFO_RESPONSE_CODE);
        curl_close($ch);
        $data = ($resp !== "" && $resp !== false) ? json_decode($resp, true) : null;
        if ($code >= 300) {
            throw new FiduciaException($code, $data);
        }
        return $data;
    }

    private static function enc(string $s): string
    {
        return rawurlencode($s);
    }

    // --- misc ---
    public function health() { return $this->request("GET", "/healthz"); }
    public function status() { return $this->request("GET", "/v1/status"); }

    // --- locks (current protocol: holder + fencing_token, key in the body) ---
    public function lockGet(string $key)
    {
        return $this->request("GET", "/v1/locks?key=" . self::enc($key));
    }
    public function lockAcquire($keys, ?string $holder = null, ?int $ttlMs = null, bool $wait = false)
    {
        return $this->request("POST", "/v1/locks/acquire",
            ["keys" => is_array($keys) ? $keys : [$keys], "holder" => $holder, "ttl_ms" => $ttlMs, "wait" => $wait]);
    }
    public function lockRelease(string $holder, int $fencingToken)
    {
        return $this->request("POST", "/v1/locks/release", ["holder" => $holder, "fencing_token" => $fencingToken]);
    }

    // --- semaphores ---
    public function semaphoreGet(string $key)
    {
        return $this->request("GET", "/v1/semaphores?key=" . self::enc($key));
    }
    public function semaphoreAcquire(string $key, int $limit, ?string $holder = null, ?int $ttlMs = null, bool $wait = false)
    {
        return $this->request("POST", "/v1/semaphores/acquire",
            ["key" => $key, "limit" => $limit, "holder" => $holder, "ttl_ms" => $ttlMs, "wait" => $wait]);
    }
    public function semaphoreRelease(string $key, string $holder, int $fencingToken)
    {
        return $this->request("POST", "/v1/semaphores/release",
            ["key" => $key, "holder" => $holder, "fencing_token" => $fencingToken]);
    }

    // --- high-level blocking / try acquisition (live-mutex style) ---

    /** tryLock: wait:false — returns a Lock if free right now, else null. */
    public function tryLock($key, int $ttlMs = 60000, ?string $holder = null): ?Lock
    {
        return $this->acquireLock(is_array($key) ? $key : [$key], false, $ttlMs, $holder, 0, 0, null);
    }

    /** lock: wait:true — blocks until acquired, the budget elapses (LockTimeout), or error. */
    public function lock($key, int $ttlMs = 60000, ?string $holder = null,
        int $maxWaitMs = 30000, int $retryIntervalMs = 250, ?int $maxRetries = null): Lock
    {
        $keys = is_array($key) ? $key : [$key];
        $got = $this->acquireLock($keys, true, $ttlMs, $holder, $maxWaitMs, $retryIntervalMs, $maxRetries);
        if ($got === null) {
            throw new LockTimeout($keys, $maxWaitMs);
        }
        return $got;
    }

    /** mustLock is an alias of lock. */
    public function mustLock($key, int $ttlMs = 60000, ?string $holder = null,
        int $maxWaitMs = 30000, int $retryIntervalMs = 250, ?int $maxRetries = null): Lock
    {
        return $this->lock($key, $ttlMs, $holder, $maxWaitMs, $retryIntervalMs, $maxRetries);
    }

    /** trySemaphore: wait:false — returns a handle if a permit is free, else null. */
    public function trySemaphore(string $key, int $limit, int $ttlMs = 60000, ?string $holder = null): ?SemaphoreHandle
    {
        return $this->acquireSemaphoreInner($key, $limit, false, $ttlMs, $holder, 0, 0, null);
    }

    /** acquireSemaphore: wait:true — blocks until a permit frees, budget elapses, or error. */
    public function acquireSemaphore(string $key, int $limit, int $ttlMs = 60000, ?string $holder = null,
        int $maxWaitMs = 30000, int $retryIntervalMs = 250, ?int $maxRetries = null): SemaphoreHandle
    {
        $got = $this->acquireSemaphoreInner($key, $limit, true, $ttlMs, $holder, $maxWaitMs, $retryIntervalMs, $maxRetries);
        if ($got === null) {
            throw new LockTimeout([$key], $maxWaitMs);
        }
        return $got;
    }

    private function acquireLock(array $keys, bool $wait, int $ttlMs, ?string $holder,
        int $maxWaitMs, int $retryIntervalMs, ?int $maxRetries): ?Lock
    {
        $holder = $holder ?? self::genHolder();
        $out = self::output($this->lockAcquire($keys, $holder, $ttlMs, $wait));
        if (!empty($out["acquired"])) {
            return new Lock($this, $keys, $holder, (int)($out["fencing_token"] ?? 0), $out["lease_expires_ms"] ?? null);
        }
        if (!$wait) {
            return null; // tryLock: held now -> fail fast
        }
        $deadline = self::nowMs() + $maxWaitMs;
        $attempts = 0;
        while ($maxRetries === null || $attempts < $maxRetries) {
            $attempts++;
            $remaining = $deadline - self::nowMs();
            if ($remaining <= 0) {
                break;
            }
            usleep(min($retryIntervalMs, $remaining) * 1000);
            $lk = $this->lockGet($keys[0])["lock"] ?? null;
            if ($lk && ($lk["holder"] ?? null) === $holder && isset($lk["fencing_token"])) {
                return new Lock($this, $keys, $holder, (int)$lk["fencing_token"], $lk["lease_expires_ms"] ?? null);
            }
        }
        return null;
    }

    private function acquireSemaphoreInner(string $key, int $limit, bool $wait, int $ttlMs, ?string $holder,
        int $maxWaitMs, int $retryIntervalMs, ?int $maxRetries): ?SemaphoreHandle
    {
        $holder = $holder ?? self::genHolder();
        $out = self::output($this->semaphoreAcquire($key, $limit, $holder, $ttlMs, $wait));
        if (!empty($out["acquired"])) {
            return new SemaphoreHandle($this, $key, $holder, (int)($out["fencing_token"] ?? 0), $out["lease_expires_ms"] ?? null);
        }
        if (!$wait) {
            return null;
        }
        $deadline = self::nowMs() + $maxWaitMs;
        $attempts = 0;
        while ($maxRetries === null || $attempts < $maxRetries) {
            $attempts++;
            $remaining = $deadline - self::nowMs();
            if ($remaining <= 0) {
                break;
            }
            usleep(min($retryIntervalMs, $remaining) * 1000);
            $sem = $this->semaphoreGet($key)["semaphore"] ?? null;
            foreach (($sem["holders"] ?? []) as $slot) {
                if (($slot["holder"] ?? null) === $holder && isset($slot["fencing_token"])) {
                    return new SemaphoreHandle($this, $key, $holder, (int)$slot["fencing_token"], $slot["lease_expires_ms"] ?? null);
                }
            }
        }
        return null;
    }

    private static function output($resp): array
    {
        $out = $resp["result"]["output"] ?? null;
        return is_array($out) ? $out : [];
    }

    private static function genHolder(): string
    {
        return "fdc-" . bin2hex(random_bytes(8));
    }

    private static function nowMs(): int
    {
        return (int)(hrtime(true) / 1000000);
    }

    // --- reader-writer locks ---
    public function rwAcquireRead(string $key, ?int $ttlMs = null, bool $wait = true)
    {
        return $this->request("POST", "/v1/rw/" . self::enc($key) . "/read", ["ttl_ms" => $ttlMs, "wait" => $wait]);
    }
    public function rwEndRead(string $key, string $lockId)
    {
        return $this->request("POST", "/v1/rw/" . self::enc($key) . "/read/end", ["lock_id" => $lockId]);
    }
    public function rwAcquireWrite(string $key, ?int $ttlMs = null, bool $wait = true)
    {
        return $this->request("POST", "/v1/rw/" . self::enc($key) . "/write", ["ttl_ms" => $ttlMs, "wait" => $wait]);
    }
    public function rwEndWrite(string $key, string $lockId)
    {
        return $this->request("POST", "/v1/rw/" . self::enc($key) . "/write/end", ["lock_id" => $lockId]);
    }

    // --- config KV ---
    public function kvGet(string $key) { return $this->request("GET", "/v1/kv/" . self::enc($key)); }
    public function kvPut(string $key, string $value, ?int $ttlMs = null)
    {
        return $this->request("PUT", "/v1/kv/" . self::enc($key), ["value" => $value, "ttl_ms" => $ttlMs]);
    }
    public function kvDelete(string $key) { return $this->request("DELETE", "/v1/kv/" . self::enc($key)); }
    public function kvList(string $prefix) { return $this->request("GET", "/v1/kv?prefix=" . self::enc($prefix)); }

    // --- leader election ---
    public function electionCampaign(string $name, string $candidate, int $ttlMs)
    {
        return $this->request("POST", "/v1/elections/" . self::enc($name) . "/campaign",
            ["candidate" => $candidate, "ttl_ms" => $ttlMs]);
    }
    public function electionRenew(string $name, string $candidate, int $fencingToken)
    {
        return $this->request("POST", "/v1/elections/" . self::enc($name) . "/renew",
            ["candidate" => $candidate, "fencing_token" => $fencingToken]);
    }
    public function electionResign(string $name, string $candidate, int $fencingToken)
    {
        return $this->request("POST", "/v1/elections/" . self::enc($name) . "/resign",
            ["candidate" => $candidate, "fencing_token" => $fencingToken]);
    }
    public function electionGet(string $name) { return $this->request("GET", "/v1/elections/" . self::enc($name)); }

    // --- service discovery ---
    public function serviceRegister(string $service, string $instanceId, string $address, int $ttlMs)
    {
        return $this->request("PUT", "/v1/services/" . self::enc($service) . "/instances/" . self::enc($instanceId),
            ["address" => $address, "ttl_ms" => $ttlMs]);
    }
    public function serviceHeartbeat(string $service, string $instanceId)
    {
        return $this->request("POST", "/v1/services/" . self::enc($service) . "/instances/" . self::enc($instanceId) . "/heartbeat");
    }
    public function serviceDeregister(string $service, string $instanceId)
    {
        return $this->request("DELETE", "/v1/services/" . self::enc($service) . "/instances/" . self::enc($instanceId));
    }
    public function serviceInstances(string $service) { return $this->request("GET", "/v1/services/" . self::enc($service)); }
    public function serviceList() { return $this->request("GET", "/v1/services"); }
}
